File: src/modules/trees/traversals/2_iterative_dfs.ts
class TreeNode {
    left: TreeNode | null
    right: TreeNode | null
    data: number
    constructor(data: number) {
        this.data = data
        this.left = null
        this.right = null
    }
}

class IterativeTreeTraversal {
    public inorderTraversal(root: TreeNode | null): void {
        if (root === null) return;
        let stack: TreeNode[] = []
        let curr: TreeNode | null = root
        while (curr !== null || stack.length > 0) {
            while (curr !== null) {
                stack.push(curr)
                curr = curr.left
            }
            curr = stack.pop()!
            console.log(curr.data)
            curr = curr.right
        }
    }
    public preorderTraversal(root: TreeNode | null): void {
        if (root === null) return;
        let stack: TreeNode[] = []
        stack.push(root)
        while (stack.length > 0) {
            let curr: TreeNode | null = stack.pop()!
            console.log(curr.data);
            if (curr.right !== null) stack.push(curr.right)
            if (curr.left !== null) stack.push(curr.left)
        }
    }
    public postorderTraversal(root: TreeNode | null): void {
        if (root === null) return;
        let stack: TreeNode[] = []
        let curr: TreeNode | null = root
        let prev: TreeNode | null = null
        while (curr !== null || stack.length > 0) {
            while (curr !== null) {
                stack.push(curr)
                curr = curr.left
            }
            curr = stack[stack.length - 1]
            if (curr.right === null || curr.right === prev) {
                console.log(curr.data)
                stack.pop()
                prev = curr
                curr = null
            } else {
                curr = curr.right
            }
        }
    }
}

File: src/modules/trees/2_traversals/2_iterative_dfs.go
package main

import "fmt"

type TreeNode struct {
	left  *TreeNode
	right *TreeNode
	data  int
}

func NewTreeNode(data int) *TreeNode {
	return &TreeNode{data: data}
}

type DfsTraversal interface {
	inorderTraversal(root *TreeNode)
	preorderTraversal(root *TreeNode)
	postorderTraversal(root *TreeNode)
}

type IterativeTreeTraversal struct{}

func (itt *IterativeTreeTraversal) inorderTraversal(root *TreeNode) {
	if root == nil {
		return
	}

	stack := []*TreeNode{}
	curr := root

	for curr != nil || len(stack) > 0 {
		for curr != nil {
			stack = append(stack, curr)
			curr = curr.left
		}
		curr = stack[len(stack)-1]
		// Pop the top item from the stack
		stack = stack[:len(stack)-1]
		fmt.Println(curr.data)
		curr = curr.right
	}
}

func (itt *IterativeTreeTraversal) preorderTraversal(root *TreeNode) {
	if root == nil {
		return
	}

	stack := []*TreeNode{}
	stack = append(stack, root)

	for len(stack) > 0 {
		curr := stack[len(stack)-1]
		// Pop the top item from the stack
		stack = stack[:len(stack)-1]
		fmt.Println(curr.data)
		if curr.right != nil {
			stack = append(stack, curr.right)
		}
		if curr.left != nil {
			stack = append(stack, curr.left)
		}
	}
}

func (itt *IterativeTreeTraversal) postorderTraversal(root *TreeNode) {
	if root == nil {
		return
	}

	stack := []*TreeNode{}
	var curr *TreeNode = root
	var prev *TreeNode

	for curr != nil || len(stack) > 0 {
		for curr != nil {
			stack = append(stack, curr)
			curr = curr.left
		}
		curr = stack[len(stack)-1]

		if curr.right == nil || curr.right == prev {
			fmt.Println(curr.data)
			stack = stack[:len(stack)-1]
			prev = curr
			curr = nil
		} else {
			curr = curr.right
		}
	}
}
